devito/arguments.py

```diff
@@ -158,198 +139,381 @@ class PtrArgument(Argument):
-        SymbolicData) might provide multiple runtime arguments.
+class ArgumentEngine(object):
+    """ Class that encapsulates the argument derivation and verification subsystem
     """
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.stencils = stencils
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
+        self.argument_mapper = self._build_argument_mapper(parameters)
+        self.arguments = filter_ordered([x for x in self.argument_mapper
+                                         if isinstance(x, Argument)],
+                                        key=lambda x: x.name)
+        self.dims = [x for x in self.argument_mapper if isinstance(x, DimensionParameter)]
+        self.offsets = {d.end_name: v for d, v in retrieve_offsets(stencils).items()}
+
+    def handle(self, **kwargs):
+        """ The main method by which the :class:`Operator` interacts with this class.
+            The arguments passed into Operator.apply() all end up in kwargs here.
+        """
 
-    def __init__(self, *args, **kwargs):
-        pass
-
-    @property
-    def rtargs(self):
-        """ Return a list of RuntimeArgument objects that this object needs.
+        user_autotune = kwargs.pop('autotune', False)
+
+        kwargs = self._offset_adjust(kwargs)
+
+        kwargs = self._extract_children_of_composites(kwargs)
+
+        values = self._derive_values(kwargs)
+
+        # The following is only being done to update the autotune flag. The actual value
+        # derivation for the dle arguments has moved inside the above _derive_values
+        # method.
+        # TODO: Refactor so this is not required
+        dim_sizes = dict([(d.name, runtime_dim_extent(d, values))
+                          for d in self.dimensions])
+        dle_arguments, dle_autotune = self._dle_arguments(dim_sizes)
+
+        assert(self._verify(values))
+
+        arguments = OrderedDict([(k.name, v) for k, v in values.items()])
+
+        return arguments, user_autotune and dle_autotune
+
+    def _offset_adjust(self, kwargs):
+        for k, v in kwargs.items():
+            if k in self.offsets:
+                kwargs[k] = v + self.offsets[k]
+        return kwargs
+
+    def _build_argument_mapper(self, parameters):
+        # Pass through SymbolicFunction
+        symbolic_functions = [x for x in parameters if isinstance(x, SymbolicFunction)]
+        dim_dep_mapper = OrderedDict()  # Mapper for dependencies between dimensions
+        tensor_arguments = []
+        for f in symbolic_functions:
+            argument = ArgumentVisitor().visit(f)
+            tensor_arguments.append(argument)
+            for i, d in enumerate(f.indices):
+                dim_dep_mapper.setdefault(d, []).append(Dependency("gets_value_from",
+                                                                   argument, param=i))
+
+        for arg in self.dle_arguments:
+            d = arg.argument
```

Comment:
is this really necessary?

Reply:
Please elaborate

Reply:
why do you need the temporary variable ``d``, used in only one place? how does it help

Reply:
Fixed